astropy/table/row.py

```diff
@@ -29,25 +30,32 @@ class Row:
     """
 
     def __init__(self, table, index):
-        self._table = table
-        self._index = operator.index(index)
+        # Get the length of the first column.  Rely on the table being well-formed
+        # at this point.
+        n = len(next(iter(table.columns.values())))
 
-        n = len(table)
         if index < -n or index >= n:
             raise IndexError('index {0} out of range for table with length {1}'
                              .format(index, len(table)))
 
-        # Ensure that the row index is positive [#8422]
-        self._index = self._index % n
-
-    def __getitem__(self, item):
-        if self._table._is_list_or_tuple_of_str(item):
-            cols = [self._table[name] for name in item]
-            out = self._table.__class__(cols, copy=False)[self._index]
+        self._table = table
 
-        else:
-            out = self._table.columns[item][self._index]
+        # Ensure that the row index is a valid index (int) and positive [#8422]
+        self._index = operator_index(index) % n
 
+    def __getitem__(self, item):
+        try:
+            # Try the most common use case of accessing a single column in the Row.
+            # Bypass the TableColumns __getitem__ since that does more testing
+            # and allows a list of tuple or str, which is not the right thing here.
+            out = OrderedDict.__getitem__(self._table.columns, item)[self._index]
```

Comment:
If you want to future-proof this (recent python guarantees ordering for a regular `dict`...), then it could be
`super(TableColumns, self._table.columns).__getitem__(item)[self._index]`

Alternatively, since `Table.__getitem__` faces a similar quandary, perhaps have a specific `TableColumns._get_column(name)` method?

(But not a big deal either way.)

Reply:
I think the direct `OrderedDict` is the fastest and simplest.  Since `table.columns` is not going to be a subclass this should be fine.

I'm quite keen to transition to the native (ordered) dict when possible!  (dict is a lot faster than OrderedDict as well).  I just put in a breadcrumb comment above the `class TableColumns(OrderedDict)` statement to not forget about this.

Reply:
OK, makes sense.